core/trino-main/src/main/java/io/trino/likematcher/LikePatternMatcher.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.likematcher;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public record LikePatternMatcher(String pattern, Optional<Character> escape, LikeMatcher likeMatcher)
+{
+    public LikePatternMatcher(String pattern, Optional<Character> escape, LikeMatcher likeMatcher)
+    {
+        this.pattern = requireNonNull(pattern, "pattern is null");
+        this.escape = requireNonNull(escape, "escape is null");
+        this.likeMatcher = requireNonNull(likeMatcher, "likeMatcher is null");
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        LikePatternMatcher that = (LikePatternMatcher) o;
+        return Objects.equals(pattern, that.pattern) && Objects.equals(escape, that.escape);
```

Comment:
Where is the comment explaining why `likeMatcher` is not part of equality?

Where is the class-level guarantee that the `likeMatcher` is indeed derived from `pattern, escape` and can be ignored in equals?

What about optimize? We cannot treat unoptimized and optimized matchers as being equivalent, can we?

Reply:
Changes here are based on what was asked here
> we should add a wrapper that contains the pattern, escape and compiled LikeMatcher and use that as the value for LikePatternType. The equality for it would need to be defined just based on pattern and escape.

@martint do you see a problem with the caching key not taking into account whether the Matcher is "optimized" or not ?

Reply:
> Changes here are based on what was asked for here
> 
> > we should add a wrapper that contains the pattern, escape and compiled LikeMatcher and use that as the value for LikePatternType. The equality for it would need to be defined just based on pattern and escape.

This did not say "remove class-level guarantees that `equals` makes sense".
I would understand it more if the wrapper had  a construction way that guarantees the assumptions we're making about it.

Reply:
"optimized" doesn't matter for the caching key. That's used only to distinguish the case where we're constructing LikeMatcher per row (e.g., for dynamic patterns) or only once for the query (when the pattern and escape are constant).